src/materialized/src/http/sql.rs

```diff
@@ -28,3 +34,46 @@ pub async fn handle_sql(
         Err(e) => Err((StatusCode::BAD_REQUEST, e.to_string())),
     }
 }
+
+#[derive(Serialize)]
+struct ErrorResponse {
+    error: String,
+}
+
+pub async fn handle_sql_ws(
+    AuthedClient(mut client): AuthedClient,
+    ws: WebSocketUpgrade,
+) -> impl IntoResponse {
+    ws.on_upgrade(|mut ws| async move {
+        if let Err(e) = run_ws(&mut ws, &mut client).await {
+            let _ = send_ws_response(
+                &mut ws,
+                ErrorResponse {
+                    error: e.to_string(),
+                },
+            );
+        }
+    })
+}
+
+async fn run_ws(ws: &mut WebSocket, client: &mut SessionClient) -> Result<(), anyhow::Error> {
```

Comment:
This is cool and great that it works. However I think this could be better. I think we could make the websocket interface an interactive client just like pgwire. Instead of only accepting one message, assuming it's a tail, and returning it, we can have this loop and take any number of queries, streaming back results along with a query completed message, just like pgwire. It'd still allow clients to connect, send a single request, and wait for results, so it doesn't seem like it'd increase the difficulty of users at all.

Reply:
That seems ideal to me too. I'd be happy to work on that.